### PR TITLE
bpo-32972: Add unittest.AsyncioTestCase

### PR DESCRIPTION
**Pull request by dave-shawley**
This PR adds co-routine support to unittest by adding a new sub-class of `unittest.TestCase` named `unittest.AsyncioTestCase`.  It extends `unittest.TestCase` by adding the following functionality:

- a new event loop is created and terminated with each test method & the *same* event loop is used throughout the entire test method execution from setup through cleanup
- asynchronous test setup and teardown are supported by new methods - `asyncSetUp` and `asyncTearDown`
- test methods that are decorated with the `async` keyword are run on the event loop.  Undecorated methods are run as normal methods.
- cleanup hooks that are decorated with the `async` keyword are run on the event loop.  Undecorated methods are run as normal methods.

I made the following conscious decisions during the implementation that _may_ be somewhat controversial so I want to mention them here (in no particular order) along with the rationale.

1. **there is no direct support for other event loop implementation**: other event loop implementations can be plugged in using [a custom EventLoopPolicy implementation](https://docs.python.org/3/library/asyncio-policy.html).  It can be inserted by overriding `setUpClass`.  I believe that other loop implementations can either plug in using the event loop policy or provide their own asynchronous test case implementation (e.g., [tornado.testing](http://www.tornadoweb.org/en/stable/testing.html)).
2. **asyncio.iscoroutine is used to detect async methods**: the two places that I need to detect whether a callable is a co-routine or not are: (a) test methods and (b) cleanup hooks added with [addCleanup](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addCleanup).  Both are explicit (opt-in) actions so they are completely controlled by the test writer.
3. **outstanding asyncio Tasks are not terminated**: I'm still on the fence about whether the test should fail if there are outstanding tasks or if it should gather them.
4. **async test case is a separate class**: I decided to use a sub-class so that it is immediately obvious that a test methods are intended to run on the event loop.  This also insulates `unittest.TestCase` from unexpected breakages and maintains it's current behavior.

**Open Questions**
1. What should happen when `len(asyncio.all_tasks())` is non-zero after cleanups are run?
2. Should `unittest.AsyncioTestCase` include a global test run timeout?

I still have to amend commits to update documentation but I wanted to have the discussion about the implementation choices before I started writing up examples and updating the existing documentation suite.

https://bugs.python.org/issue32972

### CODE REVIEW

**Comment by matrixise** [comment]
Hi,

Thanks for your PR, but this one is not related to the issue. In this PR, you add the support for asyncio but it is not the topic of the bpo issue.

Maybe you have to open an other issue and assign this PR to the issue.

**Comment by dave-shawley** [comment]
@matrixise ick ... early morning mistake.  This is associated with BPO 32972, renaming the PR

File: Lib/unittest/case.py [hunk 1]
```diff
@@ -1439,3 +1454,150 @@ def shortDescription(self):
...
+    In the latter case, the event loop IS NOT RUNNING; however, you
+    can use it to run async code if necessary.
+
+    Clean up functions will be run on the event loop if they are
+    detected as co-routines; otherwise, they are called as standard
+    functions.  After ALL cleanup calls are completed, the event loop
+    is stopped and closed.
+
+    When subclassing AsyncioTestCase, the event loop is available
+    via the ``event_loop`` property.  Tests can assume that
```
Comment (matrixise) [inline Lib/unittest/case.py]: about ``event_loop`` I am not sure, usually we found ``loop`` in the documentation of asyncio

File: Lib/unittest/case.py [hunk 2]
```diff
@@ -1439,3 +1454,150 @@ def shortDescription(self):
...
+            self.event_loop.stop()
+        self.event_loop.run_until_complete(
+            self.event_loop.shutdown_asyncgens())
+        asyncio.set_event_loop(None)
+        if not self.event_loop.is_closed():
+            self.event_loop.close()
+        self.__event_loop = None
+
+    def doCleanups(self):
+        """
```
Comment (matrixise) [inline Lib/unittest/case.py]: The docstring should start on the first line, https://www.python.org/dev/peps/pep-0008/#documentation-strings

File: Lib/unittest/case.py [hunk 3]
```diff
@@ -81,6 +82,16 @@ def testPartExecutor(self, test_case, isTest=False):
             self.success = self.success and old_success
 
 
+class _IgnoredOutcome(object):
```
Comment (matrixise) [inline Lib/unittest/case.py]: In Python3, we prefer to use the new style
https://docs.python.org/3/reference/compound_stmts.html#class-definitions
```python
# new style
class _IgnoredOutcome:
    pass
```
is equivalent to 
```python
# old style
class _IgnoredOutcome(object):
    pass
```

so in this case, use the new style and not the old one.

File: Lib/unittest/test/test_asyncio_test_case.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
+import asyncio
+import collections
+import unittest
+
+
+class Test(object):
```
Comment (matrixise) [inline Lib/unittest/test/test_asyncio_test_case.py]: See my previous comment about the construction of a class
```python
class Test:
    pass
```
is better

File: Lib/unittest/case.py [hunk 5]
```diff
@@ -1439,3 +1454,150 @@ def shortDescription(self):
...
+    async def asyncTearDown(self):
+        """Hook method for deconstructing the test fixture after testing it.
+
+        This method invokes ``tearDown`` inline so subclasses MUST invoke
+        this method using the ``await`` keyword when extending it.
+        """
+        self.tearDown()
+
+    @property
+    def event_loop(self):
```
Comment (matrixise) [inline Lib/unittest/case.py]: in the documentation, we prefer to use the `loop` variable for the reference to the event loop.

Examples:
```:meth:`loop.call_soon` ``` or ```:meth:`loop.set_debug` ``` etc...

maybe you should do the same.

File: Lib/unittest/test/test_asyncio_test_case.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    suite.run(result)
+    if result.testsRun == 0:
+        raise AssertionError(f'test {test_class} not run')
+    if should_fail and result.wasSuccessful():
+        raise AssertionError(f'test {test_class} unexpectedly succeeded')
+    elif not should_fail and not result.wasSuccessful():
+        raise AssertionError(f'test {test_class} unexpectedly failed')
+
+
+class EventLoopTracking(object):
```
Comment (matrixise) [inline Lib/unittest/test/test_asyncio_test_case.py]: just `class EventLoopTracking:`

**Review by matrixise — CHANGES_REQUESTED** [review]
Thank for your contribution but could you add a NEWS entry with the blurb tool.

`pip install blurb`

**Comment by dave-shawley** [comment]
@matrixise - thanks for the review.  I amended a few commits to address your feedback. I still intend to add some documentation in _Doc/library/unittest.rst_ and _Doc/library/asyncio.rst_ as well.  I wanted to make sure that the implementation was acceptable before starting the documentation related tasks.

File: Lib/unittest/case.py [hunk 7]
```diff
@@ -1483,3 +1486,149 @@ def shortDescription(self):
...
+        """
+        self.tearDown()
+
+    @property
+    def loop(self):
+        """Active event loop for the test case."""
+        if self.__loop is None:
+            self.__loop = asyncio.new_event_loop()
+            self.__loop.set_debug(True)
+            asyncio.set_event_loop(self.__loop)
```
Comment (PetterS) [inline Lib/unittest/case.py:1589]: Does the test case have to modify the global state like this? It is calling `self.loop.run_until_complete` when running things.

Comment (dave-shawley) [reply Lib/unittest/case.py:1589]: I think that it does.  This is the only way to ensure that the code under test gets the correct loop.

**Comment by asvetlov** [comment]
Please take a look on #13386

**Comment by asvetlov** [comment]
Implemented by #13386

**Pull request closed without merging**